src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -28050,6 +28076,375 @@ class CL1 {}
+    }
+    void Method()
+    {
+        _ = local();
+        _ = local2();
+
+        async IAsyncEnumerable<string> local()
+        {
+            yield return null; // 3
+            await Task.Delay(1);
+        }
+        async IAsyncEnumerable<string?> local2()
+        {
+            yield return null;
+            await Task.Delay(1);
+        }
+    }
+}";
+            CreateCompilationWithTasksExtensions(new[] { source, AsyncStreamsTypes }, options: WithNonNullTypesTrue()).VerifyDiagnostics(
+                // (8,22): warning CS8603: Possible null reference return.
+                //         yield return null; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "null").WithLocation(8, 22),
+                // (9,22): warning CS8603: Possible null reference return.
+                //         yield return null; // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "null").WithLocation(9, 22),
+                // (25,26): warning CS8603: Possible null reference return.
+                //             yield return null; // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "null").WithLocation(25, 26)
+                );
+        }
+
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/31057"), CompilerTrait(CompilerFeature.AsyncStreams)]
+        public void Yield_IAsyncEnumerator()
+        {
+            var source = @"
+using System.Collections.Generic;
+using System.Threading.Tasks;
+class C
+{
+    async IAsyncEnumerator<string> M()
+    {
+        yield return null; // 1
+        yield return null; // 2
+        await Task.Delay(1);
+    }
+    async IAsyncEnumerator<string?> M2()
+    {
+        yield return null;
+        yield return null;
+        await Task.Delay(1);
+    }
+    void Method()
+    {
+        _ = local();
+        _ = local2();
+
+        async IAsyncEnumerator<string> local()
+        {
+            yield return null; // 3
+            await Task.Delay(1);
+        }
+        async IAsyncEnumerator<string?> local2()
+        {
+            yield return null;
+            await Task.Delay(1);
+        }
+    }
+}";
+            CreateCompilationWithTasksExtensions(new[] { source, AsyncStreamsTypes }, options: WithNonNullTypesTrue()).VerifyDiagnostics(
+                // (8,22): warning CS8603: Possible null reference return.
+                //         yield return null; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "null").WithLocation(8, 22),
+                // (9,22): warning CS8603: Possible null reference return.
+                //         yield return null; // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "null").WithLocation(9, 22),
+                // (25,26): warning CS8603: Possible null reference return.
+                //             yield return null; // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "null").WithLocation(25, 26)
+                );
+        }
+
         [Fact]
```

Comment:
I'd expect a `yield break` positive test just to verify it generates no warnings.